src/base.c

```diff
@@ -233,20 +258,36 @@ base_block_alloc(tsdn_t *tsdn, base_t *base, extent_hooks_t *extent_hooks,
 		void *addr = (void *)block;
 		assert(((uintptr_t)addr & HUGEPAGE_MASK) == 0 &&
 		    (block_size & HUGEPAGE_MASK) == 0);
-		/* base == NULL indicates this is a new base. */
-		if (base != NULL || opt_metadata_thp == metadata_thp_always) {
-			/* Use hugepage for the new block. */
+		if (opt_metadata_thp == metadata_thp_always) {
 			pages_huge(addr, block_size);
-		}
-		if (base != NULL && base_is_single_block(base) &&
-		    opt_metadata_thp == metadata_thp_auto) {
-			/* Make the first block THP lazily. */
-			base_block_t *first_block = base->blocks;
-			assert((first_block->size & HUGEPAGE_MASK) == 0);
-			pages_huge(first_block, first_block->size);
-			if (config_stats) {
-				assert(base->n_thp == 0);
-				base->n_thp += first_block->size >> LG_HUGEPAGE;
+		} else if (opt_metadata_thp == metadata_thp_auto &&
```

Comment:
It feels like a lot of this logic belongs in base_auto_thp_switch (or, that logic belongs here or something); the decision making about what and where thps should be turned on are better if unified. Right now the fact that a0 is special is sort of threaded through implicitly in a lot of places rather than having one place where we say "here's the a0 logic and there's the non-a0 logic".

Reply:
Good point. I'll consolidate the two places.

Reply:
Just updated. Now the logic is contained in base_auto_thp_triggered and base_auto_thp_switch.